lib/Dialect/OM/Transforms/FreezePaths.cpp

```diff
@@ -0,0 +1,240 @@
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/InnerSymbolTable.h"
+#include "circt/Dialect/OM/OMAttributes.h"
+#include "circt/Dialect/OM/OMOps.h"
+#include "circt/Dialect/OM/OMPasses.h"
+
+using namespace circt;
+using namespace om;
+
+namespace {
+struct PathVisitor {
+  PathVisitor(hw::InstanceGraph &instanceGraph, hw::InnerRefNamespace &irn)
+      : instanceGraph(instanceGraph), irn(irn) {}
+  LogicalResult processPath(PathOp path);
+  LogicalResult run(Operation *op);
+  hw::InstanceGraph &instanceGraph;
+  hw::InnerRefNamespace &irn;
+};
+} // namespace
+
+static LogicalResult getAccessPath(Value value, size_t fieldId,
+                                   std::string &result) {
+  auto type = value.getType();
+  SmallString<64> name;
+  while (fieldId) {
+    if (auto aliasType = dyn_cast<hw::TypeAliasType>(type))
+      type = aliasType.getCanonicalType();
+    if (auto structType = dyn_cast<hw::StructType>(type)) {
+      auto index = structType.getIndexForFieldID(fieldId);
+      auto &element = structType.getElements()[index];
+      name += ".";
+      name += element.name.getValue();
+      type = element.type;
+      fieldId -= structType.getFieldID(index);
+    } else if (auto arrayType = dyn_cast<hw::ArrayType>(type)) {
+      auto index = arrayType.getIndexForFieldID(fieldId);
+      name += "[";
+      name += std::to_string(index);
+      name += "]";
+      type = arrayType.getElementType();
+      fieldId -= arrayType.getFieldID(index);
+    } else {
+      return emitError(value.getLoc())
+             << "can't create access path with fieldID " << fieldId
+             << " in type " << type;
+    }
+  }
+  result = name.str().str();
+  return success();
+}
+
+LogicalResult PathVisitor::processPath(PathOp path) {
+  auto *context = path->getContext();
+  auto &symbolTable = irn.symTable;
+
+  StringRef targetKind;
+  switch (path.getTargetKind()) {
+  case TargetKind::DontTouch:
+    targetKind = "OMDontTouchedReferenceTarget";
+    break;
+  case TargetKind::Reference:
+    targetKind = "OMReferenceTarget";
+    break;
+  case TargetKind::MemberReference:
+    targetKind = "OMMemberReferenceTarget";
+    break;
+  case TargetKind::MemberInstance:
+    targetKind = "OMMemberInstanceTarget";
+    break;
+  }
+
+  // Look up the associated HierPathOp.
+  auto hierPathOp =
+      symbolTable.lookup<hw::HierPathOp>(path.getTargetAttr().getAttr());
+  auto namepath = hierPathOp.getNamepathAttr().getValue();
+
+  // The name of the module which starts off the path.
+  StringAttr topModule;
+  // The path from the top module to the target. Represents a pair of instance
+  // and its target module's name.
+  SmallVector<std::pair<StringAttr, StringAttr>> modules;
+  // If we're targeting a component or port of the target module, this will hold
+  // its name.
+  StringAttr component;
+  // If we're indexing in to the component, this will be the access path.
+  std::string field;
```

Comment:
Along with above suggestion, this would be probably something like:

`SmallString<64> field`